feat: Technical review — Related website sets docs (#30500)
mdn/content: files/en-us/web/api/storage_access_api/index.md

You might have accompanying utility sites that have components embedded in all the other sites, for example to provide Single Sign-On (`sso-example.com`) or general personalization services (`services-example.com`). These utility sites will want to share state with the sites they are embedded in via cookies. They cannot share first-party cookies because they are on different domains, and third-party cookies will no longer work in browsers that block them.

In such situations, site owners often encourage users to add their site as an exception or to disable third-party cookie blocking policies entirely. Users who wish to continue to interact with their content are forced to greatly relax their blocking policy for resources loaded from all embedded origins and possibly across all websites.

The Storage Access API is intended to solve this problem; embedded cross-site content can request unrestricted access to third-party cookies on a frame-by-frame basis, via the {{domxref("Document.requestStorageAccess()")}} method. It can also check whether it already has access via the {{domxref("Document.hasStorageAccess()")}} method.

### Unpartitioned versus partitioned cookies

> It is important to note that the Storage Access API is only needed to provide access to _unpartitioned_ third-party cookies. This means cookies stored in the traditional way they have historically been stored since the early web — all cookies set on the same site are stored in the same cookie jar. This is in contrast to _partitioned_ cookies, where embedded resources under each top-level site are given a unique cookie storage space, thereby making tracking users across sites via cookies impossible.

Reviewer: ```suggestion
It is important to note that the Storage Access API is only needed to provide access to _unpartitioned_ third-party cookies. This means cookies stored in the traditional way they have historically been stored since the early web — all cookies set on the same site are stored in the same cookie jar. This is in contrast to _partitioned_ cookies, where embedded resources under each top-level site are given a unique cookie storage space, thereby making tracking users across sites via these cookies impossible.
```

Browsers have various mechanisms to partition third party cookie access, for example [Firefox Total Cookie Protection](https://blog.mozilla.org/products/firefox/firefox-rolls-out-total-cookie-protection-by-default-to-all-users-worldwide/) and [Cookies Having Independent Partitioned State (CHIPS)](/en-US/docs/Web/Privacy/Partitioned_cookies).

> When we talk about third-party cookies in the content of the Storage Access API, we implicitly mean _unpartitioned_ third-party cookies.

Reviewer: ```suggestion
When we talk about third-party cookies in the context of the Storage Access API, we implicitly mean _unpartitioned_ third-party cookies.
```

I think you meant "context" here? Although "content" also works, feel free to ignore.

Author: Yup, I meant context ;-)

   ```

6. Usage of this feature may be blocked by a {{httpheader("Permissions-Policy/storage-access", "storage-access")}} [Permissions Policy](/en-US/docs/Web/HTTP/Permissions_Policy) set on your server.

> **Note:** The document may also be required to pass additional browser-specific checks. Examples: allowlists, blocklists, on-device classification, user settings, anti-[clickjacking](/en-US/docs/Glossary/Clickjacking) heuristics, or prompting the user for explicit permission.

## Browser-specific variations

Although the API surface is the same, websites using the Storage Access API should expect differences in the level and extent of third-party cookie access they receive between different browsers, due to differences in their storage access policies.

### Chrome

- Cookies must have [`SameSite=None`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#samesitesamesite-value) explicitly set on them, because the default value for Chrome is `SameSite=Lax` (`SameSite=None` is the default in Firefox and Safari).
- Cookies must have the [`Secure`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#secure) attribute set on them.

> - The storage access grants are phased out after 30 days of browser usage passed without user interaction. Successful use of the Storage Access API extends this limit by another 30 days. This doesn't occur when {{domxref("Document.requestStorageAccessFor()")}} is called because the user is already on the page.

Reviewer: ```suggestion
- The storage access grants are phased out after 30 days of browser usage passed without user interaction. Interaction with the embedded content extends this limit by another 30 days. This doesn't occur when {{domxref("Document.requestStorageAccessFor()")}} is called because the user is already on the page.
```

- {{domxref("Document.hasStorageAccess()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with a boolean value indicating whether the document has access to third-party cookies.
- {{domxref("Document.requestStorageAccess()")}}
  - : Requests access to third-party cookies from embedded content; returns a {{jsxref("Promise")}} that resolves if the access was granted, and rejects if access was denied.
- {{domxref("Document.requestStorageAccessFor()")}} {{experimental_inline}}
  - : Allows top-level sites to request third-party cookie access on behalf of embedded content originating from another site in the same [related website set](/en-US/docs/Web/API/Storage_Access_API/Related_website_sets). Returns a {{jsxref("Promise")}} that resolves if the access was granted, and rejects if access was denied.

> **Note:** User interaction propagates to the promise returned by these methods, allowing the callers to take actions that require user interaction without requiring a second click from the user. For example, a caller could open a pop-up window from the resolved promise without triggering Firefox's pop-up blocker.

### Additions to other APIs

- {{domxref("Permissions.query()")}}

>   - : You can call `Permissions.query()` with a value of `"storage-access"` to query whether third-paty cookie access has been granted in general, i.e. to another same-site embed. If so, you can call `requestStorageAccess()` without user interaction, and the promise will resolve automatically.

Reviewer: ```suggestion
  - : You can call `Permissions.query()` with a value of `"storage-access"` to query whether third-party cookie access has been granted in general, i.e. to another same-site embed. If so, you can call `requestStorageAccess()` without user interaction, and the promise will resolve automatically.
```

>   - : You can call `Permissions.query()` with a value of `"storage-access"` to query whether third-paty cookie access has been granted in general, i.e. to another same-site embed. If so, you can call `requestStorageAccess()` without user interaction, and the promise will resolve automatically.

Reviewer: > If so, you can call `requestStorageAccess()` without user interaction, and the promise will resolve automatically.

This is true in Chrome and in Firefox AFAIK, but it's not true in Safari. Safari always requires a user gesture when calling `requestStorageAccess()`.

And furthermore, Safari doesn't support the "storage-access" permission yet, AFAIK. I believe Safari's implementation uses the old "storage access map" concept from the spec, due to their use of the per-page model.

Author: OK, I don't really want to start talking about browser support in the text, as that's the job of the browser compat data, but I have added a table for the `"storage-access"` query feature to this page in the compat section, and I have updated the wording to "In supporting browsers, ..." to give an extra hint.

I've also updated the text in the "Using..." page to explicitly state that it is Safari that doesn't support the permission, hence why the `try...catch` block is needed.

I've also added a separate entry in about the `"top-level-storage-access"` query feature.

## Examples

- See [Using the Storage Access API](/en-US/docs/Web/API/Storage_Access_API/Using) for an implementation guide with code examples.
- See [Storage Access API Demo](https://storage-access-api-demo.glitch.me/) for a live demo.

## API methods

- {{domxref("Document.hasStorageAccess()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with a boolean value indicating whether the document has access to third-party cookies.
- {{domxref("Document.requestStorageAccess()")}}
  - : Requests access to third-party cookies from embedded content; returns a {{jsxref("Promise")}} that resolves if the access was granted, and rejects if access was denied.

> - {{domxref("Document.requestStorageAccessFor()")}} {{experimental_inline}}

Author: I've had a go at addressing this by adding "a proposed extension to the Storage Access API" in the introductory descriptive text on the main landing page, the using guide, and thye method reference page. To do more would be a bit overkill, IMO: we already label it as experimental, and the repo spec is already linked in the specification section. And the Chrome-specific stuff is supposed to be about differences in the behavior of the standard SAA.
